### DISCUSSION TITLE
How to detect client connection state?

### DISCUSSION THREAD

**Discussion by nicolaiunrein** [urql-graphql/urql#1832] [Q&A] [answered]
I wondered if there is a way to detect and possibly subscribe to connection state changes.
It would be really helpful if I could fire an event whenever the client disconnects eg. clear subscription data etc. Any thoughts?

**Comment by kitten** [comment, marked as answer]
I'm not sure if you're referring to subscription transport libraries here, but if you mean, let's say, `graphql-ws`, then you'll instantiate their client first, which means that it then is responsible for describing its connection state. Last time I checked I wasn't sure whether they implemented a ping system to detect the connection state, but in either way, it's actually a more complicated problem than it sounds.

If you use WebSockets, then you'll have to note that they're built to try to retain "broken" connections for as long as possible, so they can recover on unstable internet connections. That means, depending on the network conditions, even if they drop ACK packets, as far as I know, they won't actually tell you that something's gone wrong.

In the case of Graphcache, where we account for simple network errors, we actually only react to them if we do see a network error, so you may want to do something similar. That's because there is an event that the browser provides for connection changes, but it won't tell you whether you're actually online or offline.

**Reply by nicolaiunrein** [reply]
Hey! Thank you for your detailed answer. Yes, I was referring to subscriptions and will be looking into the transport libraries.

> If you use WebSockets, then you'll have to note that they're built to try to retain "broken" connections for as long as possible, so they can recover on unstable internet connections. That means, depending on the network conditions, even if they drop ACK packets, as far as I know, they won't actually tell you that something's gone wrong.

I did not know that! I do see connection errors in the console though and think that the connection is actually closed.
